example/socks/socks_client.hpp

```diff
@@ -0,0 +1,491 @@
+﻿//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#pragma once
+
+#include "error.hpp"
+
+#include <boost/asio/spawn.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/streambuf.hpp>
+#include <boost/asio/read.hpp>
+#include <boost/asio/write.hpp>
+
+#include <iostream>
+#include <string>
+#include <memory>
+
+using boost::asio::ip::tcp;
+
+namespace socks {
+
+    template <typename Handler>
+    class socks_client
+        : public std::enable_shared_from_this<socks_client<Handler>>
+    {
+    public:
+        enum {
+            SOCKS_VERSION_4 = 4,
+            SOCKS_VERSION_5 = 5
+        };
+        enum {
+            SOCKS5_AUTH_NONE = 0x00,
+            SOCKS5_AUTH = 0x02,
+            SOCKS5_AUTH_UNACCEPTABLE = 0xFF
+        };
+        enum {
+            SOCKS_CMD_CONNECT = 0x01,
+            SOCKS_CMD_BIND = 0x02,
+            SOCKS5_CMD_UDP = 0x03
+        };
+        enum {
+            SOCKS5_ATYP_IPV4 = 0x01,
+            SOCKS5_ATYP_DOMAINNAME = 0x03,
+            SOCKS5_ATYP_IPV6 = 0x04
+        };
+        enum {
+            SOCKS5_SUCCEEDED = 0x00,
+            SOCKS5_GENERAL_SOCKS_SERVER_FAILURE,
+            SOCKS5_CONNECTION_NOT_ALLOWED_BY_RULESET,
+            SOCKS5_NETWORK_UNREACHABLE,
+            SOCKS5_CONNECTION_REFUSED,
+            SOCKS5_TTL_EXPIRED,
+            SOCKS5_COMMAND_NOT_SUPPORTED,
+            SOCKS5_ADDRESS_TYPE_NOT_SUPPORTED,
+            SOCKS5_UNASSIGNED
+        };
+        enum {
+            SOCKS4_REQUEST_GRANTED = 90,
+            SOCKS4_REQUEST_REJECTED_OR_FAILED,
+            SOCKS4_CANNOT_CONNECT_TARGET_SERVER,
+            SOCKS4_REQUEST_REJECTED_USER_NO_ALLOW,
+        };
+
+    public:
+        explicit socks_client(tcp::socket& socket,
+            const std::string& hostname, unsigned short port, int version,
+            const std::string& username, const std::string& password,
+            bool use_hostname, Handler handler)
+            : socket_(socket)
+            , hostname_(hostname)
+            , port_(port)
+            , version_(version)
+            , username_(username)
+            , password_(password)
+            , use_hostname_(use_hostname)
+            , handler_(handler)
+        {}
+
+        void handshake()
+        {
+            auto self = std::enable_shared_from_this<socks_client<Handler>>::shared_from_this();
+            switch (version_)
+            {
+            case SOCKS_VERSION_5:
+                boost::asio::spawn(socket_.get_io_context(), [this, self]
```

Comment:
The implementation of the composed operation should not use a stackful coroutine. Instead, it should be written using the stackless coroutine API that comes with Asio. There's a tutorial in the Beast docs which can guide you:
https://www.boost.org/doc/libs/1_68_0/libs/beast/doc/html/beast/using_io/writing_composed_operations.html

I can also help you get started by answering questions and give you some initial declarations, if you desire.